### PR TITLE
[DAG] Remove OneUse restriction when folding (shl (add x, c1), c2)

### PR DESCRIPTION
**Pull request by LiqinWeng**
Refer to https://github.com/llvm/llvm-project/pull/68972

### CODE REVIEW

File: llvm/test/CodeGen/ARM/add-like-or.ll [hunk 1]
```diff
@@ -249,27 +249,28 @@ entry:
 define i32 @multiuse(i32 %i, ptr %x, ptr %y) {
 ; CHECK-T1-LABEL: multiuse:
 ; CHECK-T1:       @ %bb.0: @ %entry
+; CHECK-T1-NEXT:    lsls r2, r0, #3
+; CHECK-T1-NEXT:    adds r1, r1, r2
+; CHECK-T1-NEXT:    ldr r1, [r1, #4]
```
Comment (wangpc-pp) [inline llvm/test/CodeGen/ARM/add-like-or.ll:254]: Is this a regression?

File: llvm/test/CodeGen/RISCV/rvv/vmulh-sdnode.ll [hunk 2]
```diff
@@ -7,14 +7,14 @@
...
-; CHECK-NEXT:    li a0, 42
+; CHECK-NEXT:    li a0, -85
 ; CHECK-NEXT:    vsetvli a1, zero, e8, mf2, ta, ma
-; CHECK-NEXT:    vmv.v.x v9, a0
-; CHECK-NEXT:    li a1, -85
-; CHECK-NEXT:    vmacc.vx v9, a1, v8
-; CHECK-NEXT:    vsll.vi v8, v9, 7
-; CHECK-NEXT:    vsrl.vi v9, v9, 1
-; CHECK-NEXT:    vor.vv v8, v9, v8
+; CHECK-NEXT:    vmul.vx v8, v8, a0
```
Comment (wangpc-pp) [inline llvm/test/CodeGen/RISCV/rvv/vmulh-sdnode.ll:12]: `vmv.v.x+vmacc.vx` to `vmul.vx+vadd.vx` may be a regression.

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 3]
```diff
@@ -3304,6 +3304,17 @@ X86TargetLowering::preferredShiftLegalizationStrategy(
...
 
+bool X86TargetLowering::isDesirableToCommuteWithShift(
+    const SDNode *N, CombineLevel Level) const {
+  assert((N->getOpcode() == ISD::SHL || N->getOpcode() == ISD::SRA ||
+          N->getOpcode() == ISD::SRL) &&
+         "Expected shift op");
+
+  if (!N->getOperand(0)->hasOneUse())
+    return false;
+  return true;
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:3315]: Can't we just say `return N->getOperand(0)->hasOneUse()` ?

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 4]
```diff
@@ -10035,7 +10035,7 @@ SDValue DAGCombiner::visitSHL(SDNode *N) {
   // Variant of version done on multiply, except mul by a power of 2 is turned
   // into a shift.
   if ((N0.getOpcode() == ISD::ADD || N0.getOpcode() == ISD::OR) &&
-      N0->hasOneUse() && TLI.isDesirableToCommuteWithShift(N, Level)) {
+      TLI.isDesirableToCommuteWithShift(N, Level)) {
```
Comment (goldsteinn) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:10038]: Simpler might be to make `TLI.isDesirableToComputeWithShift` return an enum s.t based on return its either:
1. Always preferable
2. Preferable if one use
3. Never preferable

Then you can just leave the 1-use check here as opposed to putting it in each backend.

**Pull request closed without merging**